server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/SpanServiceElasticsearchTest.java

```diff
@@ -173,16 +186,210 @@ public void testGetSpanFromEmptyDB() throws StoreException, InterruptedException
+        serverDescendantOfDescendant2.setAnnotations(serverAnnotations());
+
+        Span serverDescendantOfDescendant = new Span();
+        serverDescendantOfDescendant.setId("descendant");
+        serverDescendantOfDescendant.setParentId("descendant");
+        serverDescendantOfDescendant.setTraceId("root");
+        serverDescendantOfDescendant.setAnnotations(serverAnnotations());
+
+        storeAndWait(null, Arrays.asList(rootServerSpan,
+                clientDescendant,
+                clientDescendant2,
+                serverDescendantOfDescendant,
+                serverDescendantOfDescendant2),
+                SpanUniqueIdGenerator::toUnique);
+
+        Trace trace = spanService.getTraceFragment(null, "root");
+        Assert.assertEquals("root", trace.getId());
+        Assert.assertEquals(1, trace.getNodes().size());
+
+        InteractionNode rootConsumerNode = ((InteractionNode) trace.getNodes().get(0));
+        Assert.assertThat(rootConsumerNode, instanceOf(Consumer.class));
+        Assert.assertEquals(1, rootConsumerNode.getCorrelationIds().size());
+        Assert.assertEquals("root", rootConsumerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertEquals(2, rootConsumerNode.getNodes().size());
+        Assert.assertEquals(new HashSet<>(Arrays.asList("descendant", "descendant2")),
+                extractCorrelationIds(rootConsumerNode.getNodes()));
+
+        InteractionNode descendantProducerNode = ((InteractionNode) rootConsumerNode.getNodes().get(0));
+        Assert.assertThat(descendantProducerNode, instanceOf(Producer.class));
+        Assert.assertEquals(1, descendantProducerNode.getCorrelationIds().size());
+        Assert.assertEquals("descendant", descendantProducerNode.getCorrelationIds().get(0).getValue());
+        Assert.assertTrue(descendantProducerNode.getNodes().isEmpty());
+
+        InteractionNode descendantProducerNode2 = ((InteractionNode) rootConsumerNode.getNodes().get(1));
+        Assert.assertThat(descendantProducerNode2, instanceOf(Producer.class));
+        Assert.assertEquals(1, descendantProducerNode2.getCorrelationIds().size());
+        Assert.assertEquals("descendant2", descendantProducerNode2.getCorrelationIds().get(0).getValue());
+        Assert.assertTrue(descendantProducerNode2.getNodes().isEmpty());
+    }
+
+    @Test
+    public void testGetTraceFragmentClientRoot() throws StoreException, InterruptedException {
+        Span rootClientSpan = new Span();
+        rootClientSpan.setId("root");
+        rootClientSpan.setTraceId("root");
+        rootClientSpan.setName("rootClientSpan");
+        rootClientSpan.setAnnotations(clientAnnotations());
+
+        Span serverSpan = new Span();
+        serverSpan.setId("root");
+        serverSpan.setParentId("root");
+        serverSpan.setTraceId("root");
+        serverSpan.setAnnotations(serverAnnotations());
+
+        Span serverSpanDescendantRoot = new Span();
```

Comment:
Unclear which this is a server span for? If the initial rootClientSpan, then that client span can only have one corresponding server span, which is 'serverSpan'.